src/liboslcomp/typecheck.cpp

```diff
@@ -1180,58 +1156,422 @@ ASTfunction_call::typecheck_struct_constructor ()
 }
 
 
+/// <doc CandidateFunctions>
+///
+/// Score a set of polymorphic functions based on arguments & return type.
+///
+/// When choosing which varaint to use precedence is given to arguments and the
+/// return type is used as a possible tie breaker in case of any ambiguity.
+/// The highest score for an argument is given when it is an exact match.
+/// If the argument needs to be coerced into a type OSL will prefer going
+/// from any triple-types to any other triple-type before promotion of a
+/// float to a triple.  When converting one triple to another triple, OSL will
+/// prefer to coerce from one of the spatial triples (vector/point/normal) to
+/// another spatial triple before coercion from a spatial triple to color.
+///
+/// Float to int coercion is scored, but is currently a synmonym for kNoMatch
+/// as the spec does not allow implicit float to int conversion.
+///
+/// </script>
+/// ~~~~
+class CandidateFunctions {
+    enum {
+        kExactMatch     = 100,
+        kIntegralToFP   = 77,
+        kArrayMatch     = 44,
```

Comment:
I'm a little concerned about kArrayMatch, which as I understand it is when the function definition has a formal parameter array[], but the call has a specific length like array[5]. 

It seems to me that the substitution of exact array length for undefined length is *expected* and should have very high score (or low cost). 

As an example, let's say there are 3 parameters, two polymorphic varieties:

    void func (float a[], int b, int c)
    void func (float a[3], float b, float c)

and you call `float x[3]; int i; func (x, i, i);`

The score on the first one is 44+100+100==244, the score on the second is 100+77+77=254. So the second will be chosen. Does that match our intuition? I don't know, maybe it's fine? I'm just trying to find all the little things that might trip us up later. 

I believe that C++ looks for exact match first, then allows certain substitutions, but if more than one function can match with substitutions, it considers it an error (there is no ranking of substitutions like we are trying to do here). I think that's a little rigid and we want more flexibility for OSL, but I do see that they have maybe done a smart thing by not trying adjudicate that. I'm just thinking out loud here. Not advocating more rigidity. But if you wanted to argue that this whole effort is just pushing trouble farther down the road but it's going to come back to haunt us, I would certainly entertain that line of argument.

Reply:
ArrayMatch isn't really that low its just exact matches and int->float are extremely high.
The reason the number isn't extremely high is to provide some space trying to figure out how to score an init list in the future.
```
A. void func (vector v, int b, int c);
B. void func (float a[3], int b, int c);

func({1,2,3}, 1, 2);  // Should this call A, B, or is it ambiguous?
```
##
I really like the new feature of preferring spatial->spatial vs spatial->color, past that I have no preference on whether ambiguity is made more strict, but might easily break what is totally legal now:

```
A. float mix(float a, float b, float c);
B. float mix(int   a, int   b, float c);

// Currently legal (calls B), but would become ambiguous
func(1, 2, 1);
```
##
I get the ambiguity of `func(x3, i, i)` you're describing but in the cases providing literals it seems to me quite un-ambiguous what the user is trying to do.
```
A. void func (float a[], int b, int c);
B. void func (float a[3], float b, float c);

int i; float f;

float x5[5];
func (x5, i, i) Should call A.
func (x5, f, f) Should call A.
func (x5, 2, 2) Should call A.

float x3[3];
func (x3, i, i) Should call B.
func (x3, f, f) Should call B.
func (x3, 2, 2) Should call B.
```
Seems to do what I would expect.

Reply:
OK, makes sense to me. We can always revise if we find problems in practice.

I will try these on our production shader library before merging, just to be sure there aren't any problems we haven't seen yet.